presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java

```diff
@@ -639,19 +638,18 @@ public ActualProperties visitProject(ProjectNode node, List<ActualProperties> in
                 // ("ROW comparison not supported for fields with null elements", etc)
                 if (isExpression(expression)) {
                     Map<NodeRef<Expression>, Type> expressionTypes = getExpressionTypes(session, metadata, parser, types, castToExpression(expression), emptyList(), WarningCollector.NOOP);
-                    Type type = requireNonNull(expressionTypes.get(NodeRef.of(castToExpression(expression))));
                     ExpressionInterpreter optimizer = ExpressionInterpreter.expressionOptimizer(castToExpression(expression), metadata, session, expressionTypes);
                     Object value = optimizer.optimize(NoOpSymbolResolver.INSTANCE);
 
                     if (value instanceof SymbolReference) {
                         VariableReferenceExpression variable = toVariableReference(Symbol.from((SymbolReference) value), types);
                         ConstantExpression existingConstantValue = constants.get(variable);
                         if (existingConstantValue != null) {
-                            constants.put(output, new ConstantExpression(value, type));
+                            constants.put(output, new ConstantExpression(existingConstantValue.getValue(), output.getType()));
```

Comment:
Wonder for what case, `existingConstantValue.getType()` is different from `output.getType()`? The only think I can think of is something like `(1L, BIGINT)` vs `(1L, INTEGER)`. Can you come up with a test to verify that or add a comment/checkstate to explain what can/cannot happen.

Reply:
What's the motivation for this change? I'm also wondering when will this two type differ. But I think, if they do differ, the constant should have the output's type, not the expression.

Reply:
Yes, this only happens if the constant and output type is coercible such as INT/BIGINT and VARCHAR with different length. You can assign a  `VHARCHAR(5)` to variable with `VHARCHAR(10)`. In this case, output's type should always be used.

Yes, we should always use output type if the types are coercible. If not, (for example assigning `VARCHAR` to `BIGINT`), the logic here is wrong. I am wondering do we want to check if the two type is `coercible` or not here.   If input/output has different type, it will eventually be caught later by `DependencyChecker`. 

I definitely can add/update the test to cover such cases.